// @texttree/tn-quote/dist/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bookUrl = exports.DEBUG_MODE = void 0;
var DEBUG_MODE = true;
exports.DEBUG_MODE = DEBUG_MODE;
var bookUrl = {
  gen: 'unfoldingWord/hbo_uhb/raw/branch/master/01-GEN.usfm',
  exo: 'unfoldingWord/hbo_uhb/raw/branch/master/02-EXO.usfm',
  lev: 'unfoldingWord/hbo_uhb/raw/branch/master/03-LEV.usfm',
  num: 'unfoldingWord/hbo_uhb/raw/branch/master/04-NUM.usfm',
  deu: 'unfoldingWord/hbo_uhb/raw/branch/master/05-DEU.usfm',
  jos: 'unfoldingWord/hbo_uhb/raw/branch/master/06-JOS.usfm',
  jdg: 'unfoldingWord/hbo_uhb/raw/branch/master/07-JDG.usfm',
  rut: 'unfoldingWord/hbo_uhb/raw/branch/master/08-RUT.usfm',
  '1sa': 'unfoldingWord/hbo_uhb/raw/branch/master/09-1SA.usfm',
  '2sa': 'unfoldingWord/hbo_uhb/raw/branch/master/10-2SA.usfm',
  '1ki': 'unfoldingWord/hbo_uhb/raw/branch/master/11-1KI.usfm',
  '2ki': 'unfoldingWord/hbo_uhb/raw/branch/master/12-2KI.usfm',
  '1ch': 'unfoldingWord/hbo_uhb/raw/branch/master/13-1CH.usfm',
  '2ch': 'unfoldingWord/hbo_uhb/raw/branch/master/14-2CH.usfm',
  ezr: 'unfoldingWord/hbo_uhb/raw/branch/master/15-EZR.usfm',
  neh: 'unfoldingWord/hbo_uhb/raw/branch/master/16-NEH.usfm',
  est: 'unfoldingWord/hbo_uhb/raw/branch/master/17-EST.usfm',
  job: 'unfoldingWord/hbo_uhb/raw/branch/master/18-JOB.usfm',
  psa: 'unfoldingWord/hbo_uhb/raw/branch/master/19-PSA.usfm',
  pro: 'unfoldingWord/hbo_uhb/raw/branch/master/20-PRO.usfm',
  ecc: 'unfoldingWord/hbo_uhb/raw/branch/master/21-ECC.usfm',
  sng: 'unfoldingWord/hbo_uhb/raw/branch/master/22-SNG.usfm',
  isa: 'unfoldingWord/hbo_uhb/raw/branch/master/23-ISA.usfm',
  jer: 'unfoldingWord/hbo_uhb/raw/branch/master/24-JER.usfm',
  lam: 'unfoldingWord/hbo_uhb/raw/branch/master/25-LAM.usfm',
  ezk: 'unfoldingWord/hbo_uhb/raw/branch/master/26-EZK.usfm',
  dan: 'unfoldingWord/hbo_uhb/raw/branch/master/27-DAN.usfm',
  hos: 'unfoldingWord/hbo_uhb/raw/branch/master/28-HOS.usfm',
  jol: 'unfoldingWord/hbo_uhb/raw/branch/master/29-JOL.usfm',
  amo: 'unfoldingWord/hbo_uhb/raw/branch/master/30-AMO.usfm',
  oba: 'unfoldingWord/hbo_uhb/raw/branch/master/31-OBA.usfm',
  jon: 'unfoldingWord/hbo_uhb/raw/branch/master/32-JON.usfm',
  mic: 'unfoldingWord/hbo_uhb/raw/branch/master/33-MIC.usfm',
  nam: 'unfoldingWord/hbo_uhb/raw/branch/master/34-NAM.usfm',
  hab: 'unfoldingWord/hbo_uhb/raw/branch/master/35-HAB.usfm',
  zep: 'unfoldingWord/hbo_uhb/raw/branch/master/36-ZEP.usfm',
  hag: 'unfoldingWord/hbo_uhb/raw/branch/master/37-HAG.usfm',
  zec: 'unfoldingWord/hbo_uhb/raw/branch/master/38-ZEC.usfm',
  mal: 'unfoldingWord/hbo_uhb/raw/branch/master/39-MAL.usfm',
  mat: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/41-MAT.usfm',
  mrk: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/42-MRK.usfm',
  luk: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/43-LUK.usfm',
  jhn: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/44-JHN.usfm',
  act: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/45-ACT.usfm',
  rom: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/46-ROM.usfm',
  '1co': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/47-1CO.usfm',
  '2co': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/48-2CO.usfm',
  gal: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/49-GAL.usfm',
  eph: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/50-EPH.usfm',
  php: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/51-PHP.usfm',
  col: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/52-COL.usfm',
  '1th': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/53-1TH.usfm',
  '2th': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/54-2TH.usfm',
  '1ti': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/55-1TI.usfm',
  '2ti': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/56-2TI.usfm',
  tit: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/57-TIT.usfm',
  phm: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/58-PHM.usfm',
  heb: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/59-HEB.usfm',
  jas: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/60-JAS.usfm',
  '1pe': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/61-1PE.usfm',
  '2pe': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/62-2PE.usfm',
  '1jn': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/63-1JN.usfm',
  '2jn': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/64-2JN.usfm',
  '3jn': 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/65-3JN.usfm',
  jud: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/66-JUD.usfm',
  rev: 'unfoldingWord/el-x-koine_ugnt/raw/branch/master/67-REV.usfm'
};
exports.bookUrl = bookUrl;
