// @texttree/tn-quote/dist/components/useSelection/useSelection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _srrcl = require("../../utils/srrcl");
var _selections = require("../../utils/selections");
function useSelection(_ref) {
  var _ref$greekVerseObject = _ref.greekVerseObjects,
    greekVerseObjects = _ref$greekVerseObject === void 0 ? [] : _ref$greekVerseObject,
    _ref$targetVerseObjec = _ref.targetVerseObjects,
    targetVerseObjects = _ref$targetVerseObjec === void 0 ? [] : _ref$targetVerseObjec,
    _ref$quote = _ref.quote,
    quote = _ref$quote === void 0 ? '' : _ref$quote,
    _ref$occurrence = _ref.occurrence,
    occurrence = _ref$occurrence === void 0 ? 0 : _ref$occurrence,
    chapter = _ref.chapter,
    verses = _ref.verses;
  if (occurrence === 0 || greekVerseObjects.length === 0 || targetVerseObjects.length === 0 || quote === '' || verses.length === 0) {
    return;
  }
  var selections = (0, _srrcl.selectionsFromQuoteAndVerseObjects)({
    quote: quote,
    verseObjects: greekVerseObjects,
    occurrence: occurrence,
    chapter: chapter,
    verses: verses
  });
  var result = targetVerseObjects.map(function (el) {
    return (0, _selections.parseVerseObjects)(el, selections, {
      chapter: chapter,
      verse: verses[0]
    });
  });
  return (0, _selections.formatToString)(result);
}
var _default = useSelection;
exports["default"] = _default;
