// tsv-groupdata-parser/lib/groupsIndexParser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getArticleCategory = exports.generateGroupsIndex = void 0;

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _pathExtra = _interopRequireDefault(require("path-extra"));

var _resourcesHelpers = require("./helpers/resourcesHelpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * make sure this groupId is in a category
 * @param {String} groupId
 * @param {String} groupName - localized group name
 * @param {Object} categorizedGroupsIndex
 * @param {String} categoryName - category that should contain this groupId
 */
const addGroupToCategory = (groupId, groupName, categorizedGroupsIndex, categoryName) => {
  const groupIndexItem = getGroupIndex(groupId, groupName); // Only add the groupIndexItem if it isn't already in the category's groups index.

  if (!categorizedGroupsIndex[categoryName].some(e => e.id === groupIndexItem.id)) {
    categorizedGroupsIndex[categoryName].push(groupIndexItem); // adding group Index Item
  }
};

const generateGroupsIndex = (tnCategoriesPath, taCategoriesPath) => {
  const categorizedGroupsIndex = {
    discourse: [],
    numbers: [],
    figures: [],
    culture: [],
    grammar: [],
    other: []
  };
  const errors = [];

  const isDirectory = item => _fsExtra.default.lstatSync(_pathExtra.default.join(tnCategoriesPath, item)).isDirectory();

  const categories = _fsExtra.default.readdirSync(tnCategoriesPath).filter(isDirectory);

  categories.forEach(categoryName => {
    const booksPath = _pathExtra.default.join(tnCategoriesPath, categoryName, 'groups');

    const books = _fsExtra.default.readdirSync(booksPath).filter(item => item !== '.DS_Store');

    books.forEach(bookid => {
      const groupDataPath = _pathExtra.default.join(booksPath, bookid);

      const groupDataFiles = _fsExtra.default.readdirSync(groupDataPath).filter(filename => _pathExtra.default.extname(filename) === '.json');

      let taArticleCategory;
      let groupName;
      let contextId;
      groupDataFiles.forEach(groupDataFile => {
        try {
          const filePath = _pathExtra.default.join(groupDataPath, groupDataFile);

          const groupId = groupDataFile.replace('.json', '');

          const groupData = _fsExtra.default.readJsonSync(filePath);

          taArticleCategory = null;
          groupName = null;
          let foundLocalization = false;
          contextId = null;

          if (groupData.length > 0) {
            for (let i = 0; i < groupData.length; i++) {
              contextId = groupData[i].contextId;
              taArticleCategory = null;

              try {
                taArticleCategory = getArticleCategory(contextId.occurrenceNote, groupId);
              } catch (e) {
                // invalid occurrence note
                console.warn(`generateGroupsIndex() - Error parsing Link in Occurrence Note '${contextId.occurrenceNote}' at contextID: ${JSON.stringify(contextId)}`);
              }

              try {
                if (!taArticleCategory) {
                  console.log(`generateGroupsIndex() - Link in Occurrence Note '${contextId.occurrenceNote}' does not have category for check at contextID: ${JSON.stringify(contextId)}`);
                  taArticleCategory = 'translate';
                }

                const fileName = groupId + '.md';

                const articlePath = _pathExtra.default.join(taCategoriesPath, taArticleCategory, fileName);

                groupName = (0, _resourcesHelpers.getGroupName)(articlePath);
                addGroupToCategory(groupId, groupName, categorizedGroupsIndex, categoryName);
                foundLocalization = true;
                break; // we got the category, so don't need to search anymore
              } catch (e) {
                let message = `error finding group name: groupId: '${groupId}', index: '${i}' in bookId '${bookid}, taArticleCategory: ${taArticleCategory}, contextID: ${JSON.stringify(contextId)}' `;
                console.error('generateGroupsIndex() - ' + message, e);
              }
            }

            if (!foundLocalization) {
              addGroupToCategory(groupId, groupId, categorizedGroupsIndex, categoryName); // add entry even though we could not find localized description

              console.error(`Could not find localization for '${groupId}' in '${categoryName}', adding stub entry`);
            }
          }
        } catch (e) {
          let message = `error processing entry: bookid: ${bookid}, groupDataFile: ${groupDataFile}, taArticleCategory: ${taArticleCategory}, groupName: ${groupName}, contextID: ${JSON.stringify(contextId)}: `;
          console.error('generateGroupsIndex() - ' + message, e);
          errors.push(message + e.toString());
        }
      });
    });
  });

  if (errors.length) {
    throw new Error(`generateGroupsIndex() - error processing index:\n${errors.join('\n')}`);
  }

  return categorizedGroupsIndex;
};
/**
 * Gets the category of the given groupId from the links in the occurrenceNote
 * @param {string} occurrenceNote
 * @param {string} groupId
 * @returns {string} the matched category
 */


exports.generateGroupsIndex = generateGroupsIndex;

const getArticleCategory = (occurrenceNote, groupId) => {
  if (occurrenceNote && groupId) {
    const pattern = '(?<=rc:\\/\\/[^\\/]+\\/ta\\/man\\/)[^\\/]+?(?=\\/' + groupId + ')';
    const categoryRE = new RegExp(pattern);
    const match = occurrenceNote.match(categoryRE);

    if (match) {
      return match[0];
    }
  }

  return null;
};

exports.getArticleCategory = getArticleCategory;

const getGroupIndex = (groupId, groupName) => ({
  id: groupId,
  name: groupName
});
